### DISCUSSION TITLE
ChainUsage plot is empty while I see alignment

### DISCUSSION THREAD

**Discussion by misrak** [milaboratory/mixcr#1685] [Q&A] [answered]
Hello All,

I would like to understand why I don't see anything in the chainUsage plot while I see more than 2% of alignment.

Method:
Immediately after extraction, 5,000 nuclei per sample were loaded onto a Chromium Single Cell 3′ Chip (10X Genomics) and processed for the single-nucleus cDNA library preparation (Chromium Single Cell 3’ version 3, 10X Genomics) according to the manufacturer’s instructions without modifications. The remaining nuclei were stored at −80 °C for future applications. Before sequencing at high resolution, each cDNA library was tested using the High Sensitivity DNA Assay on the Agilent 2100 Bioanalyzer. Before sequencing, samples were also pooled into groups of 8 and tested using MiSeq to estimate sequencing parameters and mapping to the human genome. All samples were judged suitable for deep sequencing. 50,000 reads per nucleus were sequenced using the Illumina HiSeq platform. 

Commands used:
```
mixcr analyze 10x-sc-5gex --species hsa -MrefineTagsAndSort.whitelists.CELL=builtin:3M-febrary-2018 --tag-pattern "^(CELL:N{16})(UMI:N{12})\^(R2:*)" ~/beefe/fastqfiles/scRNAseq/chronicLesions/SRR15235936_1.fastq.gz ~/beefe/fastqfiles/scRNAseq/chronicLesions/SRR15235936_2.fastq.gz SRR15235936_result
```

![alignQC](https://github.com/milaboratory/mixcr/assets/13136697/6a9087c0-63ef-44eb-925a-bed27f23b20c)
![chainUsageQC](https://github.com/milaboratory/mixcr/assets/13136697/3ede26c1-84b5-4513-89d2-33db9cb7b229)

**Comment by mizraelson** [comment, marked as answer]
Hi, can you share the reports from the mixcr command? My guess is that due to various reasons there were no clones assembled from the initially aligned reads (perhaps low quality nucleotides, or other thresholds etc.) and the `.clns` file is actually empty thats why you don't see anything.

If you try running the command bellow you will most likely see the distribution of chains for alignments:

```
mixcr exportQc chainUsage SRR15235936_result.vdjca chains.pdf
```

**Comment by misrak** [comment]
When I ran the above command this is what I see:

![image](https://github.com/milaboratory/mixcr/assets/13136697/a264ae60-593a-44f2-80f9-93e1a2d69261)


align.report.txt

Input file(s): SRR15235936_1.fastq.gz,SRR15235936_2.fastq.gz
Output file(s): SRR15235936_result.vdjca
Version: 4.5.0; built=Fri Sep 22 14:39:05 CEST 2023; rev=cdb24b4fb7; lib=repseqio.v3.0.1
Command line arguments: align --report SRR15235936_result.align.report.txt --json-report SRR15235936_result.align.report.json --preset 10x-sc-5gex --sa
Analysis time: 236.05m
Total sequencing reads: 244167324
Successfully aligned reads: 6957416 (2.85%)
Coverage (percent of successfully aligned):
  CDR3: 0%
  FR3_TO_FR4: 0%
  CDR2_TO_FR4: 0%
  FR2_TO_FR4: 0%
  CDR1_TO_FR4: 0%
  VDJRegion: 0%
Alignment failed: no hits (not TCR/IG?): 233264182 (95.53%)
Alignment failed: low total score: 3945726 (1.62%)
Overlapped: 0 (0%)
Overlapped and aligned: 0 (0%)
Overlapped and not aligned: 0 (0%)
Alignment-aided overlaps, percent of overlapped and aligned: 0 (NaN%)
No CDR3 parts alignments, percent of successfully aligned: 6928063 (99.58%)
Partial aligned reads, percent of successfully aligned: 29333 (0.42%)
Realigned with forced non-floating bound: 0 (0%)
Realigned with forced non-floating right bound in left read: 0 (0%)
Realigned with forced non-floating left bound in right read: 0 (0%)
TRA chains: 5174 (0.07%)
TRA non-functional: 0 (0%)
TRB chains: 192367 (2.76%)
TRB non-functional: 0 (0%)
TRD chains: 616 (0.01%)
TRD non-functional: 0 (0%)
TRG chains: 5970182 (85.81%)
TRG non-functional: 0 (0%)
IGH chains: 151694 (2.18%)
IGH non-functional: 0 (0%)
TRAD chains: 268068 (3.85%)
TRAD non-functional: 0 (0%)
IGK chains: 100000 (1.44%)
IGK non-functional: 0 (0%)
IGL chains: 269315 (3.87%)
IGL non-functional: 1 (0%)
Trimming report:
  R1 reads trimmed left: 367566 (0.15%)
  R1 reads trimmed right: 485173 (0.2%)
  Average R1 nucleotides trimmed left: 0.06610461521051031
  Average R1 nucleotides trimmed right: 0.1069940423313973
Tag parsing report:
  Execution time: 0ns
  Total reads: 244167324
  Matched reads: 244167324 (100%)
  Projection +R1 +R2: 244167324 (100%)
  For variant 0:
    For projection +R1 +R2:
      CELL:Left position: 0
      UMI:Left position: 16
      CELL:Right position: 16
      UMI:Right position: 28
      R2:Left position: 0
      Variants: 0
      Cost: 0
      CELL length: 16
      UMI length: 12
      R2 length: 91

**Reply by mizraelson** [reply]
So, on the plot you see the distribution of alignments by chain type. Can you please also share the assemble report?

**Reply by misrak** [reply]
So the number of core alignments is 3 as it is too less it won't report it?

Assemble report:

```
Analysis date: Thu Jun 06 14:57:19 CEST 2024
Input file(s): SRR15235936_result.extended.vdjca
Output file(s): SRR15235936_result.clna
Version: 4.5.0; built=Fri Sep 22 14:39:05 CEST 2023; rev=cdb24b4fb7; lib=repseqio.v3.0.1
Command line arguments: assemble --report SRR15235936_result.assemble.report.txt --json-report SRR15235936_result.assemble.report.json SRR15235936_result.extended.vdjca SRR15235936_result.clna
Analysis time: 5.06m
Final clonotype count: 0
Reads used in clonotypes, percent of total: 0 (0%)
Average number of reads per clonotype: NaN
Reads dropped due to the lack of a clone sequence, percent of total: 6834691 (2.8%)
Reads dropped due to a too short clonal sequence, percent of total: 0 (0%)
Reads dropped due to low quality, percent of total: 0 (0%)
Reads dropped due to failed mapping, percent of total: 7 (0%)
Reads dropped with low quality clones, percent of total: 0 (0%)
Aligned reads processed: 7
Reads used in clonotypes before clustering, percent of total: 0 (0%)
Number of reads used as a core, percent of used: 0 (NaN%)
Mapped low quality reads, percent of used: 0 (NaN%)
Reads clustered in PCR error correction, percent of used: 0 (NaN%)
Reads pre-clustered due to the similar VJC-lists, percent of used: 0 (NaN%)
Clonotypes dropped as low quality: 0
Clonotypes eliminated by PCR error correction: 0
Clonotypes pre-clustered due to the similar VJC-lists: 0
Clones dropped in post filtering: 0 (NaN%)
Reads dropped in post filtering: 0.0 (0%)
Alignments filtered by tag prefix: 0 (0%)
Pre-clone assembler report:
  Number of input groups: 126401
  Number of input groups with no assembling feature: 126398
  Number of input alignments: 6834694
  Number of alignments with assembling feature: 3 (0%)
  Number of output pre-clones: 3
  Number of pre-clonotypes per group:   1
  Number of assembling feature sequences in groups with zero pre-clonotypes: 0
  Number of dropped pre-clones by tag suffix conflict: 0
  Number of dropped alignments by tag suffix conflict: 0
  Number of core alignments: 3 (0%)
  Discarded core alignments: 0 (0%)
  Empirically assigned alignments: 3 (0%)
  Empirical assignment conflicts: 0 (0%)
  Tag+VJ-gene empirically assigned alignments: 2 (0%)
  VJ-gene empirically assigned alignments: 1 (0%)
  Tag empirically assigned alignments: 0 (0%)
  Number of ambiguous groups: 0
  Number of ambiguous tag+V/J-gene combinations: 0
  Ignored non-productive alignments: 0 (0%)
  Unassigned alignments: 132 (0%)
======================================

```

**Reply by mizraelson** [reply]
`Final clonotype count: 0`. Thats why the plot is empty. `Number of core alignments: 3` is a part of Pre-clone assembler. In the latter clonotype assemble no clones are assembled it due to various reasons: low quality, do not cover CDR3 region...

**Reply by misrak** [reply]
Thank you so much for your answer. 

I wanted to ask if the alignment rate is calculated according to each nuclei or an average of each nuclei? It is because if I want to understand what is alignment of each TR* in a nuclei should I divide according to the number of each nuclei?

**Reply by mizraelson** [reply]
The alignment rate reflect the total number of reads in a FASTQ file. Roughly speaking you can divide it by the number of cells/nuclei. But you can't be entirely sure that each cell was sequenced with the same depth.

**Reply by misrak** [reply]
> `Final clonotype count: 0`. Thats why the plot is empty. `Number of core alignments: 3` is a part of Pre-clone assembler. In the latter clonotype assemble no clones are assembled it due to various reasons: low quality, do not cover CDR3 region...

When you say low quality, I was wondering what are the parameters taken into consideration by mixcr to assess that it is low quality? Is there a publication that I can read to get more information as to how the tool works when it is assessing the quality?
Thank you so much.

**Reply by mizraelson** [reply]
Well, the main issue here is rather the 3' approach. With 3' cDNA library it is very unllikely to cover CDR3 region as it is located closer to the 5'.  That is why 10x VDJ kit is 5'. The majority of reads you "lose" because of that. (`Reads dropped due to the lack of a clone sequence, percent of total: 6834691 (2.8%)`)

As for the sequencing quality, MiXCR uses several steps with sophisticated error corrections. One of these steps targets sequencing errors and relies on the Phred quality. It is called mapping. Before assembly, all alignments are divided into three groups based on nucleotide quality:
​
A) Alignments where the Phred quality of all nucleotides is above the 15.
B) Alignments with any nucleotide having a quality lower than 15.
C) Alignments where more than 70% of nucleotides in the assembling feature are below 15. This group of alignments is filtered out right away.
​
Initially, the group of high-quality alignments (A) is used for core clonotype assembly, while the low-quality alignments (B) are set aside for later processing. After building the core clonotypes, MiXCR tries to map the deferred reads. This step is essential for recovering quantitative information from low-quality reads. Each deferred record is mapped to the already assembled clonotypes. If a fuzzy match is found, the record is aggregated with the corresponding clonotype. Deferred reads with no match among the hight quality clones are dropped. This process helps correct sequencing errors and eliminate artificial diversity significantly more efficient than only filtering out the group C.

In your case:
There were no reads in group A (`Number of reads used as a core, percent of used: 0 (NaN%)`) and 7 reads in group B, but because there were no high quality clones MiXCR couldn't map them anywhere and dropped (`Reads dropped due to failed mapping, percent of total: 7 (0%)`)

**Reply by misrak** [reply]
Thank you so much for this explanation. This helps a lot! 👍
